FEAT: New doppler_redshift equivalency

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to formally expose the equivalency for conversion between Doppler redshift and radial velocity.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

* Follow up of https://github.com/astropy/astropy/pull/10185/files#r416867984
* Fixes #10240
* Could be used by spacetelescope/jdaviz#1031 (cc @kecnry)

- [x] Drop commit from #12711 once point is proven (whether RTD builds or not).

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
`81.50% (-10.26%) compared to 2984fa2` ? But how? I added tests! 🤯

nstarman commented:
> `81.50% (-10.26%) compared to 2984fa2` ? But how? I added tests! 🤯

2 tests have coverage: one with minimal dependencies, one with full. Since both are uploaded to codecov you're seeing the coverage from the minimal dependency test before the other test's coverage is uploaded.
The coverage is good now that codecov has synthesized the two reports.

mhvk commented on astropy/units/equivalencies.py:
```diff
@@ -503,9 +503,31 @@ def from_vel_en(x):
+
+    def convert_z_to_rv(z):
+        zponesq = (1 + z) ** 2
+        return (C_KMS * (zponesq - 1) / (zponesq + 1))
```
Maybe remove the outer set of parentheses?

mhvk commented on astropy/units/equivalencies.py:
```diff
@@ -503,9 +503,31 @@ def from_vel_en(x):
+
+    def convert_rv_to_z(rv):
+        beta = rv / C_KMS
+        return (np.sqrt((1 + beta) / (1 - beta)) - 1)
```
Also here the outer parentheses seem not needed.

pllim commented:
@nstarman , is there a way to properly reference `astropy.cosmology.units.redshift` in docs?

`WARNING: py:obj reference target not found: astropy.cosmology.units.redshift`

mhvk commented on astropy/units/equivalencies.py:
```diff
@@ -617,19 +639,19 @@ def convert_K_to_Jy(x_K):
         return Equivalency([(astrophys.Jy, si.K, convert_Jy_to_K, convert_K_to_Jy),
-                            (astrophys.Jy/astrophys.beam, si.K, convert_Jy_to_K, convert_K_to_Jy),],
-                           "brightness_temperature", {'frequency': frequency, 'beam_area': beam_area})
+                            (astrophys.Jy/astrophys.beam, si.K, convert_Jy_to_K, convert_K_to_Jy), ],  # noqa: E501
```
Not sure why one wouldn't just remove the `,` before the `]` instead of adding a space, since it is a list.

nstarman commented:
Lol. When I saw this PR I immediately added https://github.com/astropy/astropy/projects/7#card-75852886, but the more recent commits cover this 🎉 . Thanks!

nstarman commented:
> @nstarman , is there a way to properly reference `astropy.cosmology.units.redshift` in docs?

Hmm. It's in ``__all__``, so there shouldn't be a problem...
@mhvk, does the ``namespace`` argument in ``def_unit`` throw off RTD? That's the only thing I can think of.

mhvk reviewed (approved):
Looks very good! Only total nitpicks, which you should feel free to ignore.

mhvk commented:
On the doc link: I think for the link to work, you need to ensure there actually is a sphinx item for the unit - it being in `__all__` is not good enough, since by default sphinx only shows functions and classes, not instances.

Now in principle, I would have hoped that generating the unit summary would add links, but I think it does not (i.e., linking to `astropy.units.m` should fail too). We could adjust that like was done for the physical types... But perhaps easier is just to rephrase, writing something like "dimensionless or redshift units (see `astropy.cosmology.units`)"

nstarman commented:
See #12711 for a possible fix. While ``redshift`` is in ``cosmology.units.__all__``, ``units` is not in ``cosmology.__all__``.

nstarman commented on astropy/units/equivalencies.py:
```diff
@@ -503,9 +503,31 @@ def from_vel_en(x):
+
+def doppler_redshift():
+    """
+    Returns the equivalence between Doppler redshift (unitless
+    or `astropy.cosmology.units.redshift`) and radial velocity.
```
`astropy.cosmology.units.redshift` is not actually a Doppler redshift, but a cosmological redshift (https://en.wikipedia.org/wiki/Redshift). However, when interpreting a cosmological redshift as a recessional velocity, the math is the same. I would just change the phrasing to something like

```suggestion
    Returns the equivalence between Doppler redshift and radial velocity.
    For quantities with units of `astropy.cosmology.units.redshift`, this is also
    the equivalence between cosmological redshift and recessional velocity.
```

nstarman reviewed (changes requested):
Looks great! Just one note about terminology. Don't want folks mistaking cosmological redshift for a relativistic or gravitational one.
And thanks for extending the equivalency to the cosmology unit!

nstarman commented on astropy/units/equivalencies.py:
```diff
@@ -503,9 +503,31 @@ def from_vel_en(x):
+    rv_unit = si.km / si.s
+    C_KMS = _si.c.to_value(rv_unit)
+
+    def convert_z_to_rv(z):
+        zponesq = (1 + z) ** 2
+        return (C_KMS * (zponesq - 1) / (zponesq + 1))
```
```suggestion
    def convert_z_to_rv(z):
        if hasattr(z, "unit") and z.unit == redshift:
            raise UnitError("cannot interpret a cosmological redshift as a Doppler velocity.")
        zponesq = (1 + z) ** 2
        return (C_KMS * (zponesq - 1) / (zponesq + 1))
```

pllim replied:
This might be a little overkill. If I simply omit `redshift` altogether from the list, it raises `UnitConversionError` like any other incompatible unit. I already have a commit ready to go. Will push it out in a few mins.

mhvk replied:
This change is incorrect - the conversion functions get values in the equivalency unit, not quantity, so no need to check for units, etc.

mhvk replied:
In other words, @pllim is right, just removing `redshift` from the equivalency is good enough.

nstarman commented on astropy/units/equivalencies.py:
```diff
@@ -503,9 +503,31 @@ def from_vel_en(x):
+        beta = rv / C_KMS
+        return (np.sqrt((1 + beta) / (1 - beta)) - 1)
+
+    return Equivalency([(dimensionless_unscaled, rv_unit, convert_z_to_rv, convert_rv_to_z),
+                        (redshift, rv_unit, convert_z_to_rv, convert_rv_to_z),
+                        ], "doppler_redshift")
```
```suggestion
    return Equivalency([(dimensionless_unscaled, rv_unit, convert_z_to_rv, convert_rv_to_z),
                        ], "doppler_redshift")
```

nstarman commented on docs/units/equivalencies.rst:
```diff
@@ -502,6 +502,32 @@ In a special relativity context it can be convenient to use the
+    >>> z = 0.1 * u.dimensionless_unscaled
+    >>> z.to(u.km / u.s, u.doppler_redshift())  # doctest: +FLOAT_CMP
+    <Quantity 28487.0661448 km / s>
+
+It also takes the `~astropy.cosmology.units.redshift` unit::
+
+    >>> from astropy.cosmology import units as cu
+    >>> z = 0.1 * cu.redshift
+    >>> z.to(u.km / u.s, u.doppler_redshift())  # doctest: +FLOAT_CMP
+    <Quantity 28487.0661448 km / s>
```
```suggestion

```

nstarman commented on astropy/units/tests/test_equivalencies.py:
```diff
@@ -144,6 +144,20 @@ def test_bad_restfreqs(function, value):
+    from astropy.cosmology.units import redshift
+    for z_unit in (u.dimensionless_unscaled, redshift):
+        z_in = z * z_unit
+        rv_out = z_in.to(u.km / u.s, u.doppler_redshift())
```
This should error for cosmological redshift.

nstarman commented on astropy/units/tests/test_equivalencies.py:
```diff
@@ -144,6 +144,20 @@ def test_bad_restfreqs(function, value):
+    for z_unit in (u.dimensionless_unscaled, redshift):
+        z_in = z * z_unit
+        rv_out = z_in.to(u.km / u.s, u.doppler_redshift())
+        z_out = rv_out.to(z_unit, u.doppler_redshift())
```
But this doesn't have to. It's interpreting a doppler velocity as a redshift velocity, which is fine.
OTOH, one might want the equivalency to error, for simplicity.

pllim replied:
I agree. OK one way but not the other way is too confusing.

nstarman reviewed (commented):
Rolling back cosmological redshift support...
Sorry again.

pllim commented:
@nstarman , thanks for the review and sorry about the redshift confusion. Cosmology wasn't my strongest subject. I rolled back its support. I also cherry-picked a commit from #12711 to see if it fixes the RTD here. I will drop that cherry-pick once it is proven good or otherwise.

@mhvk , thanks for the review! Yes, I couldn't help myself to a little PEP 8 fixes in that file... Hopefully I have addressed all your comments as well.

mhvk reviewed (approved):
Looks great!  And sorry that I also did not catch the fact that one really should not convert cosmological redshift to velocity! Duh! Thanksfully, @nstarman did catch it.

pllim commented:
Hmm looks like RTD being slow again. I'll check tomorrow.

pllim commented:
@nstarman , unless you have other ideas on fixing the docs you want me to try, I will rephrase as @mhvk suggested in https://github.com/astropy/astropy/pull/12709#issuecomment-1009369868 tomorrow. Thanks!

pllim commented:
CI passed. @nstarman , do you wish to re-review since you requested changes? Thanks!

nstarman reviewed (approved):
Looks great! Thanks @pllim for fixing the Cosmology equivalency.
I've tested that the following correctly fails.
```python
z = 0.1 * cu.redshift
z.to(u.km / u.s, u.doppler_redshift())
```

pllim commented:
Thanks for the reviews!
